### PR TITLE
Mark never-written variables as CONSTANT in the analyzer

### PR DESCRIPTION
**Pull request by garretfick**
## Summary

Case 3 of #1612. A variable with an initializer that the program never writes holds its initial value for the whole run. A new analyzer transform, `xform_mark_unwritten_constants`, gives such declarations the `CONSTANT` qualifier at the end of type resolution, so the semantic rules and code generation see one notion of a constant variable. Code generation can then fold reads of it, starting with `LEN` of a constant string, without reasoning about writes itself. The `LEN` fold is a follow-up; this PR only establishes the qualifier.

## Stack

Targets `main`. The prerequisites have merged: #1654 (the resolver classifies user-typed initializers) and #1653 (P4009 collects globals only). The branch is rebased onto `main` after #1653, so the diff shows only the transform.

## Changes since the last review

- **Writes resolve to one declaration.** A write is resolved to the declaration the name reaches where it occurs: the symbol environment for the enclosing units (a method's declaration over its block's), `callee_resolution` for the block in the `EXTENDS` chain that declares an inherited field and for an instance's type, and the global through `VAR_EXTERNAL`. Only that declaration is blocked. A same-named declaration in another unit is unaffected. Only a target the transform cannot resolve to one declaration (a member through an array element or nested field, a configuration path, a constructor argument) blocks every declaration of the name.
- **The type environment is the one authority on whether a type is a function block.** The transform holds no table of its own.
- **`has_initial_value` and `input_expr` are methods on the DSL types**, not bare functions in the transform.
- **The design doc** states that the transform is an optimization a non-optimized or debug build may leave out, and that gating it is one call in `stages::resolve_types`.
- The by-name workaround for the P4009 rule is gone; #1653 fixes the rule instead.

## Design

`specs/design/constant-variable-inference.md` records the decisions and carries `REQ-CVI-analyzer-*` requirements with conformance tests.

- **Every construct that can change a variable counts**: assignment to any part of it, `FOR` control variables, `=>` outputs, `VAR_IN_OUT` arguments, `REF`/`ADR`/`REF=`, instance and `VAR_CONFIG` member initializers, program connection sinks, writable `VAR_ACCESS` paths, and SFC action associations. A callee whose parameter directions cannot be resolved is treated as writing every variable argument.
- **`VAR_INPUT` arguments stay reads**, so `LEN(msg)` and `timer(PT := delay)` leave `msg` and `delay` constant.
- **Only declarations the existing `CONSTANT` rules accept are marked**, so P4008, P4009 and P4010 are never introduced. Globals and their externals are marked together. Structures, references, located, retained and parameter declarations are left alone.

## Contract for later stages

After `resolve_types`, a `CONSTANT` qualifier means "holds its initial value for the whole run" whether it came from source or from this transform. A build that must let the operator change variables should run without the transform.

## Verification

`cd compiler && just` passes on the rebased branch (build, coverage, clippy, fmt, duplication check). Unit tests cover the corner cases; the conformance tests pin each requirement.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01PyMa6n6EDcNK6178uDdPUt

### CODE REVIEW

File: specs/design/constant-variable-inference.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+After `stages::resolve_types` returns, a `VarDecl` whose qualifier is
+`DeclarationQualifier::Constant` holds its initial value for the whole
+run, whether the qualifier came from the source or from this transform.
+Code generation may read the initializer instead of the variable's storage
+and may skip the storage altogether when nothing else needs it.
+
+Two consequences follow, and both are deliberate:
+
+- A tool that lets the operator change a variable while the program runs
```
Comment (garretfick) [inline specs/design/constant-variable-inference.md]: This is true, but we could also run the code generation without this step. That's what a debug/no-optimized build would do. This isn't an all or nothing. We could configure this behavior.

File: compiler/analyzer/src/xform_mark_unwritten_constants.rs [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+}
+
+/// Returns whether `init` is an initializer that `rule_var_decl_const_initialized`
+/// accepts on a `CONSTANT` declaration without any further check.
+///
+/// Exhaustive on purpose: a new initializer kind must decide here whether a
+/// never-written variable of that kind may be marked.
+fn has_constant_initializer(init: &InitialValueAssignmentKind) -> bool {
+    match init {
+        InitialValueAssignmentKind::Simple(si) => si.initial_value.is_some(),
```
Comment (garretfick) [inline compiler/analyzer/src/xform_mark_unwritten_constants.rs]: This is a bare function. Should this be part of InitialValueAssignmentKind's impl block? Same for the next function.

File: compiler/analyzer/src/xform_mark_unwritten_constants.rs [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+
+/// Gathers the name of every variable the library can write, and the global
+/// and external declarations the marking has to keep consistent.
+struct WriteCollector<'a> {
+    function_blocks: &'a FunctionBlocks<'a>,
+    type_environment: &'a TypeEnvironment,
+    function_environment: &'a FunctionEnvironment,
+    /// The function-block instances of the unit being walked.
+    instances: InstanceTypes,
+    written: HashSet<Id>,
```
Comment (garretfick) [inline compiler/analyzer/src/xform_mark_unwritten_constants.rs]: Something isn't right with this. `Id` has scope. I think this is saying we don't understand scopes for determining what's written.

File: compiler/analyzer/src/xform_mark_unwritten_constants.rs [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+        match record {
+            SymbolicVariableKind::Named(named) => self.instances.type_of(&named.name).is_some(),
+            _ => true,
+        }
+    }
+
+    /// Whether `type_name` names a function block, declared in the library
+    /// or supplied by the standard library.
+    fn is_function_block(&self, type_name: &TypeName) -> bool {
+        self.function_blocks.contains(type_name)
```
Comment (garretfick) [inline compiler/analyzer/src/xform_mark_unwritten_constants.rs]: Why would we need both here? This sounds like we don't trust the type environment.

File: specs/design/constant-variable-inference.md (hunk 1 above)
Comment (garretfick) [reply specs/design/constant-variable-inference.md]: Agreed. Reworded this section: the transform is an optimization a non-optimized or debug build leaves out, so a build that lets the operator change variables runs without it and never reads a folded value that disagrees with storage. Gating it is one call in `stages::resolve_types`; the doc says that is not wired to a build option yet.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: compiler/analyzer/src/xform_mark_unwritten_constants.rs (hunk 2 above)
Comment (garretfick) [reply compiler/analyzer/src/xform_mark_unwritten_constants.rs]: Moved: `InitialValueAssignmentKind::has_initial_value()` in the DSL says whether a declaration states a value; the transform layers only its own policy on top (structures and references stay unmarked, for the reasons in the design doc). The argument helper is `ParamAssignmentKind::input_expr()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: compiler/analyzer/src/xform_mark_unwritten_constants.rs (hunk 3 above)
Comment (garretfick) [reply compiler/analyzer/src/xform_mark_unwritten_constants.rs]: You're right, it was by name. Writes now resolve to the declaration they reach: the symbol environment answers which scope a name reaches from the current one (method over block over global, `VAR_EXTERNAL` to its global), `callee_resolution` answers which block in the `EXTENDS` chain declares an inherited field and what type an instance has for `inst.member` writes. The written set is keyed by `(ScopeKind, Id)` and only that declaration is blocked; a same-named variable in another unit is unaffected. Only a target that does not resolve to one declaration (a member through an array element or nested field, a `VAR_CONFIG` path, a constructor argument) blocks every declaration of the name. REQ-021 and its test now say this.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: compiler/analyzer/src/xform_mark_unwritten_constants.rs (hunk 4 above)
Comment (garretfick) [reply compiler/analyzer/src/xform_mark_unwritten_constants.rs]: It was distrust, and it was covering for a declaration that was mis-shaped at the source: the type resolver left `inst : FB := (x := 1)` as a structure initializer, so consumers had to guess. #1654 fixes that in the resolver, which also removes the predicate #1649 added to `InstanceTypes::declare`. Here the type environment alone answers whether a type is a function block; the library lookup is only for the declaration's parameters, which the type environment does not hold.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: compiler/analyzer/src/xform_mark_unwritten_constants.rs [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+                    self.mark_member(&fb_type, field);
+                }
+            }
+            SymbolicVariableKind::SelfRef(_) => self.mark(field),
+            _ => self.mark_any_scope(field),
+        }
+    }
+
+    /// Marks every member an instance initializer sets, at any depth.
+    fn mark_member_inits(&mut self, fb_type: &TypeName, inits: &[StructureElementInit]) {
```
Comment (garretfick) [inline compiler/analyzer/src/xform_mark_unwritten_constants.rs:241]: I'm unclear why we need to mark member inits. If we initialized and never changed, wouldn't that be a constant?
